### DISCUSSION TITLE
Licenses/Copyright in llama.cpp / ggml / whisper.cpp

### DISCUSSION THREAD

**Discussion by ikawrakow** [ggml-org/llama.cpp#6394] [General]
@ggerganov 

I see the following copyright notice in `ggml-sycl.cpp`
```
//
// MIT license
// Copyright (C) 2024 Intel Corporation
// SPDX-License-Identifier: MIT
//

//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
```

I would like to also explicitly state the copyright on the code that I have contributed. What is the best way to accomplish this?

**Comment by ggerganov** [comment]
I don't know what is the common practice in this kind of projects, so open to suggestions. I'm fine with adding contributors to the copyrights of the project as long as they agree to the terms of the MIT license. The main goal is the code to be completely free to be used, modified and distributed by anyone. I personally do not plan to apply any sort of restrictions on the code that I've written for this project. If there is better way to express this than the existing license, let me know.

My expectation is that people contributing to this project will share this sentiment. I don't want to be managing who owns which part of the code or the ways that they allow or disallow modifications. The reason is that I want to have general control of the architecture and structure of the code base and be able to modify and move things around as I find appropriate. If I had to consider for each piece of code by which person was contributed and what restrictions they put on it, this would complicate the development process to be worth it.

**Comment by Jeximo** [comment]
If there's a copyright restriction, then it's not **REALLY** a contribution, it's an obligation.

The copyright change came from this suggestion: https://github.com/ggerganov/llama.cpp/pull/5780#discussion_r1513252698

It's better to **delete** the [PR](https://github.com/ggerganov/llama.cpp/pull/5780#discussion_r1513252698) rather than get tangled in copyright.

**Comment by ikawrakow** [comment]
I'm not a lawyer, so don't take as a legal advice, but the way I understand it the situation is this:
1. @ggerganov did not ask contributors to assign copyright to him
2. @ggerganov did not ask contributors to specify a license under which they are contributing

The consequence of 1. is that the copyright on large portions of this code is not owned by @ggerganov as the LICENSE stipulates
```
MIT License

Copyright (c) 2023 Georgi Gerganov
...
```

The consequence of 2. is that large portions of this code base do not have a license. Hence, strictly speaking, no-one is allowed to do anything with the code in this repository. 

But more to my original point: @ggerganov I don't think your reply did answer my question. I asked what would be the best way to have  my copyright on the code I have contributed specified. That would allow me to do so, and then allow me to specify a license. I personally prefer other open source licenses to MIT.

**Reply by ggerganov** [reply]
I already stated that all the code in this project will remain under MIT license. There is no option for code that wants to be part of this project to be copyrighted under other licenses. I have simply not provided such mechanism currently and I am not planning to do so

**Reply by slaren** [reply]
I am not a lawyer either, but as I understand, contributors retain the copyright to their work, which they implicitly license as MIT when they contribute to this project, according to the `LICENSE` file. This is actually part of the [github terms of service](https://github.com/github/site-policy/blob/main/Policies/github-terms/github-terms-of-service.md#6-contributions-under-repository-license). But maybe the `LICENSE` file should be updated to reflect the fact that there isn't a single copyright holder of the entire project.

**Comment by ikawrakow** [comment]
@ggerganov 
> I already stated that all the code in this project will remain under MIT license.

This is a clear statement unlike "My expectation is that people contributing to this project will share this sentiment." that you stated earlier. People sharing the sentiment of freely sharing the code does not require a MIT license and can be achieve under various other licenses.

@slaren
You may not be a lawyer, but you are clearly a much better one than myself :-). According to the GitHub ToS that you kindly referred  me to, I have agreed to license all of my contribution under the MIT license (although I did not actually explicitly agree to that, instead, my agreement was obtained by me having a GitHub account and as a result agreeing to the GitHub ToS, which I had not read until today).

@ggerganov 
The result of this is that you need to include a copyright notice with my name in the LICENSE file of this repository, and all other repositories to which you may have copied my contributions.

I feel a cleaner solution may be for you to remove all my contributions from all of your repositories, but I do understand that you don't need to do so.

**Reply by JohannesGaessler** [reply]
>According to the GitHub ToS that you kindly referred me to, I have agreed to license all of my contribution under the MIT license (although I did not actually explicitly agree to that, instead, my agreement was obtained by me having a GitHub account and as a result agreeing to the GitHub ToS, which I had not read until today).

Even without the Github ToS the branch via which you submitted your contributions had a LICENSE file that clearly expressed the terms of the MIT license. I'm not going to pretend like I'm a lawyer myself but I really do not think you have a legal leg to stand on here. If you did every single major repository on Github would at some point be copyright trolled because it would be extremely lucrative for people to make some contributions and then demand license fees after the code has been merged and is in production use (and this isn't happening).

>The result of this is that you need to include a copyright notice with my name in the LICENSE file of this repository, and all other repositories to which you may have copied my contributions.

The next time you submit a PR you can edit the license file to include an additional copyright notice. If Georgi merges your PR he will be legally obligated to keep the copyright notice.

>I feel a cleaner solution may be for you to remove all my contributions from all of your repositories, but I do understand that you don't need to do so.

This is not a clean solution at all. The source code, even with all of your contributions removed, has become a derived work of those contributions. It would be necessary to remove not just your contributions but all "inspired" contributions from other people as well.

**Reply by ggerganov** [reply]
> The result of this is that you need to include a copyright notice with my name in the LICENSE file of this repository, and all other repositories to which you may have copied my contributions.

In #6405 I propose adding an `AUTHORS` file that lists all contributors to the project.

> I have agreed to license all of my contribution under the MIT license (although I did not actually explicitly agree to that, instead, my agreement was obtained by me having a GitHub account and as a result agreeing to the GitHub ToS, which I had not read until today).

I'm open to suggestions how to make the terms of new contributions more clear to everyone. We don't need to change anything that is already implied by the existing ToS - just find a way to improve clarity in order to avoid this kind of confusion. I see that other projects typically have `CONTRIBUTING` files - we should look into creating one.

> I feel a cleaner solution may be for you to remove all my contributions from all of your repositories, but I do understand that you don't need to do so.

The negative impact from such action to all downstream projects that rely on `llama.cpp` will be quite big, so I consider that it would be very irresponsible by me to do that.

@ikawrakow Based on this request, I assume that you are no longer planning to contribute to `llama.cpp`. Hence, I've revoked your collaborator access from the project

**Comment by ikawrakow** [comment]
@JohannesGaessler 

> This is not a clean solution at all. The source code, even with all of your contributions removed, has become a derived work of those contributions. It would be necessary to remove not just your contributions but all "inspired" contributions from other people as well.

Not quite. I'm actually not sure what you are arguing about (and your attempt to label me a troll is not appreciated). I did state that Georgi has no legal obligation to remove my contributions but, if he did, this does not include derived or "inspired" work as per MIT license. 

>  The next time you submit a PR you can edit the license file to include an additional copyright notice. If Georgi merges your PR he will be legally obligated to keep the copyright notice.

Also not quite. I do own the copyright on about 25k LOC, give or take, and I did not assign the copyright to Georgi, so he is indeed obliged to add my name to the copyright notice as per the license that he selected. Perhaps you yourself feel differently, and are prepared to assign the copyright of your contributions to Georgi, or feel content with your name not appearing in the LICENSE file, but I do not.

Why do I think that removing my contributions may be the cleaner solution? As an example, let's look at `ggml-sycl.cpp` that is one trigger for this discussion. In that file I observe a large amount of code that is nearly a verbatim copy of the code I have contributed, and therefore own the copyright, and yet I do not see my name appearing in the copyright notice. Now, the general consensus around the Internet appears to be that in the case of the MIT license it is not necessary to include a copyright notice in each file, one LICENSE file at the root of the repository is sufficient. But we cannot have it both ways. Either there is a single LICENSE file and Intel removes their copyright notice from `ggml-sycl.cpp`, or there are copyright notices in all files as appropriate. Perhaps you feel differently, and are content with Intel copying your code nearly verbatim without adding your name to the copyright notice, but I do not.  If Georgi removed i- and k-quants, that discussion goes away. I can give you other examples, but I'm on the go and writing this on my phone, so it is kind of tedious.

**Reply by JohannesGaessler** [reply]
>and your attempt to label me a troll is not appreciated

It is not my intent to call you a troll. I only intended to argue about the legal situation. In this context "[copyright troll](https://en.wikipedia.org/wiki/Copyright_troll)" is an established term. My intent was to argue that if your contributions had not actually been licensed as MIT (intentional or not) then copyright trolls would have already picked this up as a strategy.

**Reply by darkacorn** [reply]
either way .. complaining about that months after the fact .. he should have thought about that first

**Comment by Jeximo** [comment]
Is it possible to **remove** *Intel Copyright* from `ggml-sycl.cpp`?

**Comment by darkacorn** [comment]
i think at this point its the best option to remove your code from the project all together - to save the hassle - id love to see the legislation to that (that there is a legal obligation to mention you IN THE CODE/ copyright notice)

its commonly understood that when you contribute to a project - you get mentioned as contributor - this is sheer trolling - appreciate it or not - just don't contribute if you go down that route after months of working on a project

**Reply by darkacorn** [reply]
also .. ggerganov .. maybe have a lawyer prep a waiver to get all rights to the contributions going forward- even if its commonly understood to be the case anyway - bigger companies do that (i know this is a MIT side project) + its a hassle but if 1 start - you can assume that this will happen more often - its for the better of the whole project if a contributor has  to sign that first

i assume / hope most would not mind as they still get mentioned as a contributor anyway - but id love to hear / read opinions on that

**Comment by ikawrakow** [comment]
@ggerganov 

> The negative impact from such action to all downstream projects that rely on llama.cpp will be quite big, so I consider that it would be very irresponsible by me to do that.

Thanks for acknowledging that. 

> @ikawrakow Based on this request, I assume that you are no longer planning to contribute to llama.cpp. Hence, I've revoked your collaborator access from the project

I did not actually say that.

> In https://github.com/ggerganov/llama.cpp/pull/6405 I propose adding an AUTHORS file that lists all contributors to the project.

Isn't it slightly more complicated than that? I'm not a lawyer, hence I'm not really sure if you can handle the copyright situation that way without my agreement in this repository. But I do find my code in the [ggml](https://github.com/ggerganov/ggml) and [whisper.cpp](https://github.com/ggerganov/whisper.cpp) repositories, where I never contributed. It ended up there by you copying it, which you can of course do, but only after adding a copyright notice as per MIT license.

**Reply by JohannesGaessler** [reply]
>Isn't it slightly more complicated than that? I'm not a lawyer, hence I'm not really sure if you can handle the copyright situation that way without my agreement in this repository. But I do find my code in the [ggml](https://github.com/ggerganov/ggml) and [whisper.cpp](https://github.com/ggerganov/whisper.cpp) repositories, where I never contributed. It ended up there by you copying it, which you can of course do, but only after adding a copyright notice as per MIT license.

The exact wording of the MIT license is:

>The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

It only refers to the "above" copyright notice which as of right now is "Copyright (c) 2023 Georgi Gerganov". There is no obligation to include any other copyright notice until this is changed on master.

In any case, unrelated to any copyright notices I think it would make sense to add comments like "this function has been adapted from function X in file Y, originally written by Z" to the code. The information provided by git does not tell you who wrote the original code but this is likely who you would want to contact for specific issues. Also you could check the implementation on one backend vs. the implementation on another.

**Reply by rolandtannous** [reply]
no need @JohannesGaessler 

There is an element of foreseeability under the law.
When someone contributes code under an MIT license, regardless of the fact they have failed to read the applicable GitHub TOS (not really an excuse), the law expects them to reasonably foresee that the code they have contributed to an opensource project subject to an MIT license, will be subject to copy and reuse.  If its deemed necessary to include a notice when another repo reuses the code, even though copyright notices aren't mandatory acc to the Berne convention,  check the suggestion under paragraph 6 of my reply post here for a workaround:
https://github.com/ggerganov/llama.cpp/discussions/6394#discussioncomment-8964214

**Comment by rolandtannous** [comment]
Hello everyone,

This GitHub thread made its way to discord and I personally became aware of it yesterday.

I came here to share some preliminary perspective as someone with over 12 years of experience in commercial dispute resolution (ADR).I have also previously provided opinions and assisted in resolving two software licensing disagreements last year. However, please note that I am not a copyright lawyer, and the following should not be construed as legal advice. If you require professional legal guidance, please consult with your own lawyer.

I want to start by acknowledging that many open-source projects on GitHub have had similar discussions, so it's important not to take this matter personally or too much to heart.

With that said, @ikawrakow , I kindly request that you tone down your approach and acknowledge that some of your initial claims regarding the applicable license to your code contributions were unintentionally false. You just have not read the applicable GitHub Terms of Service. All the while, throughout this thread, @ggerganov has demonstrated good intentions and remained open to discussion. 


To keep the conversation constructive and to help resolve this issue, please consider the following notes:

 


**1- Copyrightable work:**

  Not all works or contributions entitle one to a copyright. 
  The work has to be substantive, among other criteria, in order to be copyrightable. Typo fixes, and trivial fixes are non-substantive and therefore non copyrightable.
  
Imagine what a mess it would be if every one line contributor starts claiming he’s entitled to a copyright. 
Things would go out of hand and more importantly hinder the progress of this open source project.

Having said that, @ikawrakow  appears to have contributed substantively (correct me if I am wrong), and everyone should be aware of whether the law considers copyright notices to be mandatory or not (refer to next paragraph).

 
**2- Copyright notices:**

Pursuant to the Berne convention, copyright notices are not mandatory for contributors to retain ownership of their copyright. 

The United States, for example, joined the Berne Convention after the MIT and BSD licenses were already in use, which explains why these licenses still contain explicit language about reproducing copyright notices.

However, it is important to note that copyright notices are optional, and there is no legal obligation to include them. This means that contributors retain their copyright ownership even without including explicit notices.

If people still wish to include copyright notices, please refer to paragraph 6 "Moving Forward" below for a practical workaround that doesn’t hinder the progress of the project.



**3- Project License**

This project uses an MIT license, and as per the GitHub terms of service, which @slaren correctly highlighted, contributors agree to license their contributions to the project under the same MIT license by using GitHub. This agreement is in effect regardless of whether a contributor has read the GitHub TOS beforehand.

Failing to read the GitHub TOS does not exempt a contributor from this agreement or imply that they did not expressly agree to the terms. Such an argument is not legally valid. By using GitHub, contributors have already agreed to the terms of service.
Furthermore, contributors cannot typically retract the code they have already contributed under the license that was in effect at the time of their contribution, even if they had not read the TOS previously. Doing so may potentially result in a breach of the GitHub TOS.



**4- Code reuse in closed source commercial**

If someone doesn’t want their contributed code to be used in closed source commercial projects, they should stop contributing to projects that don’t use a GPLv3 or AGPLv3 license.

This project clearly uses an MIT license. 


5- **Code contributed directly to repo vs code copied from other works**

Code that is contributed directly to a repository falls under the licensing scheme prescribed in the GitHub Terms of Service which is applicable to inbound code unto a repo that has an existing license in force. However, code that is copied from another repository, which has its own existing licensing, may have different attribution requirements. 

For example, Intel's SYCL code might have specific attribution requirements that differ from the GitHub TOS.
It is crucial to recognize that these two scenarios are not identical and cannot be compared as such at all levels.

**6- Moving forward**

If @ggerganov  decides to attribute or include a copyright notice, despite it not being mandatory (as discussed in paragraph 2), a suitable approach worth considering, is the approach suggested  by Ben Balter, the head of engineering operations at GitHub. In his blog post (https://ben.balter.com/2015/06/03/copyright-notices-for-websites-and-open-source-projects/), Balter suggests including a notice that reads:

> Copyright [year project started] - [current year], [project founder] and the [project name] contributors.

Since the list of contributors' names is publicly available under the project's repository, this approach is considered more than sufficient. By following this format, the project can acknowledge the contributions of all individuals involved without the need to list each contributor's name explicitly in the notice.


Also In the future, 
Might be a good idea to consider having a contributor license agreement moving forward. There is a GitHub integration for that here:
https://github.com/cla-assistant/cla-assistant

Also be aware that a developer certificate of origin is an option albeit not necessary and an CLA is sometimes in itself enough. 
https://developercertificate.org/
 Here is a GitHub integration for a DCO:
https://github.com/probot/dco   Hope this helps

Hope this helps,
Roland

**Reply by ggerganov** [reply]
Thank you - this is quite useful.

Based on this and to the best of my understanding of the situation, I don't see any real conflict as it stands. According to pt.2, there seems to be no need to explicitly list names in the copyright notice, since everyone who has contributed has been already attributed copyright of their work implicitly by the Github ToS. In addition to that, I would like to point out that the README of `llama.cpp` clearly states that this project is the primary place for developing features for the `ggml` library. In turn, the `ggml` README also clearly states that large portion of the development of the library is currently occurring in the `llama.cpp` and `whisper.cpp` projects. Whenever possible, contributions across all three repositories are carried over via the corresponding git commits provided by the contributors (via `git-am`, see the sync scripts), which preserves authorship.
 
I understand that we are discussing the situation and nobody here is a lawyer so nothing stated so far should be taken with absolute certainty, so I remain open to any further recommendations of how to improve the license/copyright state of the projects moving forward.

**Comment by jart** [comment]
@ikawrakow Your name is now recorded in the llamafile codebase. Your name will also be included inside every llamafile binary. https://github.com/Mozilla-Ocho/llamafile/commit/9fc233cb3c6b91e5e41d7503f2e87a73e67c5eb6 Let me know if I can do anything more to help your efforts be recognized.

**Reply by ikawrakow** [reply]
Thank you, @jart, this is very kind.

**Reply by jart** [reply]
You earned it!
